packages/web-components/src/utils/states.ts

```diff
@@ -0,0 +1,161 @@
+import { CustomStatesSetSupported } from './element-internals.js';
+
+export type ValidationStateValue =
+  | 'bad-input'
+  | 'custom-error'
+  | 'has-message'
+  | 'pattern-mismatch'
+  | 'range-overflow'
+  | 'range-underflow'
+  | 'step-mismatch'
+  | 'too-long'
+  | 'too-short'
+  | 'type-mismatch'
+  | 'valid'
+  | 'value-missing';
+
+export type AvatarStateValue =
+  | 'beige'
+  | 'blue'
+  | 'brass'
+  | 'brown'
+  | 'cornflower'
+  | 'cranberry'
+  | 'dark-green'
+  | 'dark-red'
+  | 'forest'
+  | 'gold'
+  | 'grape'
+  | 'lavender'
+  | 'light-teal'
+  | 'lilac'
+  | 'magenta'
+  | 'marigold'
+  | 'mink'
+  | 'navy'
+  | 'peach'
+  | 'pink'
+  | 'platinum'
+  | 'plum'
+  | 'pumpkin'
+  | 'purple'
+  | 'red'
+  | 'royal-blue'
+  | 'seafoam'
+  | 'steel'
+  | 'teal';
+
+export type StateValue =
+  | 'align-end'
+  | 'align-start'
+  | 'anchor'
+  | 'auto-resize'
+  | 'block'
+  | 'bold'
+  | 'bordered'
+  | 'brand'
+  | 'center'
+  | 'checked'
+  | 'circular'
+  | 'danger'
+  | 'disabled'
+  | 'display-shadow'
+  | 'dot'
+  | 'end'
+  | 'error'
+  | 'expanded'
+  | 'extra-large'
+  | 'extra-small'
+  | 'filled-darker'
+  | 'filled-lighter'
+  | 'fit-center'
+  | 'fit-contain'
+  | 'fit-cover'
+  | 'fit-none'
+  | 'focus-visible'
+  | 'ghost'
+  | 'horizontal'
+  | 'huge'
+  | 'icon-only'
+  | 'icon'
+  | 'important'
+  | 'indeterminate'
+  | 'informative'
+  | 'inline'
+  | 'inset'
+  | 'inverted'
+  | 'italic'
+  | 'justify'
+  | 'large'
+  | 'medium'
+  | 'monospace'
+  | 'multi-line'
+  | 'neutral'
+  | 'nowrap'
+  | 'numeric'
+  | 'outline'
+  | 'pressed'
+  | 'primary'
+  | 'resize-both'
+  | 'resize-horizontal'
+  | 'resize-vertical'
+  | 'resize'
+  | 'rounded'
+  | 'semibold'
+  | 'severe'
+  | 'shadowed'
+  | 'single-line'
+  | 'size-100'
+  | 'size-200'
+  | 'size-400'
+  | 'size-500'
+  | 'size-600'
+  | 'size-700'
+  | 'size-800'
+  | 'size-900'
+  | 'size-1000'
+  | 'small'
+  | 'square'
+  | 'strikethrough'
+  | 'strong'
+  | 'submenu'
+  | 'subtle'
+  | 'success'
+  | 'tint'
+  | 'tiny'
+  | 'transparent'
+  | 'truncate'
+  | 'underline'
+  | 'user-invalid'
+  | 'user-valid'
+  | 'vertical'
+  | 'warning'
+  | AvatarStateValue
+  | ValidationStateValue;
+
+/**
+ * A cache for the state values.
+ *
+ * @internal
+ */
+const statesMap = new Map<string, string>();
+
+/**
+ * Returns a string that represents a state.
+ *
+ * @param state - the state value to convert to a string.
+ * @returns a string that represents a state.
+ *
+ * @public
+ */
+export function state(state: StateValue): string {
+  if (!statesMap.has(state)) {
+    if (!CustomStatesSetSupported) {
+      statesMap.set(state, `[state--${state}]`);
+    } else {
+      statesMap.set(state, `:state(${state})`);
+    }
+  }
+
+  return statesMap.get(state)!;
```

Comment:
This looks good. I like the map. Do we have any idea of the performance tradeoffs of writing to the `statesMap` whenever the component first mounts versus memory overhead of importing a const as string and echoing it in the style template? Is it one of those situations where we save on the bundle, pay a first time cost, but make it up over n-components?

Reply:
Thanks for providing the jsperf test offline - I updated it to show the potential difference between the `stateSelector` function and using `css.partial`: https://jsperf.app/hisegi/6

This run updates the function to store `css.partial` tagged templates instead of raw template literals: https://jsperf.app/hisegi/8

It looks like the performance difference is the same in both cases, at 3x faster for `stateSelector`.